simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java

```diff
@@ -107,34 +110,109 @@ protected static boolean shouldUseCompression(HttpExchange exchange) {
+            return new CustomThreadFactory(Executors.defaultThreadFactory(), daemon);
+        }
+    }
+
+    protected final HttpServer server;
 
 
     /**
      * Start a HTTP server serving Prometheus metrics from the given registry.
      */
-    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry) throws IOException {
+    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry, boolean daemon) throws IOException {
         server = HttpServer.create();
         server.bind(addr, 3);
         HttpHandler mHandler = new HTTPMetricHandler(registry);
         server.createContext("/", mHandler);
         server.createContext("/metrics", mHandler);
-        server.setExecutor(Executors.newFixedThreadPool(5));
-        server.start();
+        server.setExecutor(Executors.newFixedThreadPool(5, CustomThreadFactory.defaultThreadFactory(daemon)));
+        start(daemon);
+    }
+
+    /**
+     * Start a HTTP server serving Prometheus metrics from the given registry using non-daemon threads.
+     */
+    public HTTPServer(InetSocketAddress addr, CollectorRegistry registry) throws IOException {
+        this(addr, registry, false);
     }
 
     /**
      * Start a HTTP server serving the default Prometheus registry.
      */
+    public HTTPServer(int port, boolean daemon) throws IOException {
+        this(new InetSocketAddress(port), CollectorRegistry.defaultRegistry, daemon);
+    }
+
+    /**
+     * Start a HTTP server serving the default Prometheus registry using non-daemon threads.
+     */
     public HTTPServer(int port) throws IOException {
-        this(new InetSocketAddress(port), CollectorRegistry.defaultRegistry);
+        this(port, false);
     }
 
     /**
      * Start a HTTP server serving the default Prometheus registry.
      */
+    public HTTPServer(String host, int port, boolean daemon) throws IOException {
+        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry, daemon);
+    }
+
+    /**
+     * Start a HTTP server serving the default Prometheus registry using non-daemon threads.
+     */
     public HTTPServer(String host, int port) throws IOException {
-        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry);
+        this(new InetSocketAddress(host, port), CollectorRegistry.defaultRegistry, false);
+    }
+
+    /**
+     * Start a HTTP server by making sure that its background thread inherit proper daemon flag.
+     */
+    private void start(boolean daemon) {
+        if (daemon == Thread.currentThread().isDaemon()) {
+            server.start();
+        } else {
+            FutureTask<Void> startTask = new FutureTask<Void>(new Runnable() {
+                @Override
+                public void run() {
+                    server.start();
+                }
+            }, null);
+            CustomThreadFactory.defaultThreadFactory(daemon).newThread(startTask).start();
+            try {
```

Comment:
Is this really necessary? We've just re-throwing the exception.

Reply:
Otherwise i have to implement internal class which should implement Runnable and store exceptions in the instance fields to re-throw them later in the main thread:
```java
class StartTask implements Runnable {
    private volatile RuntimeException exception;
    private volatile Error error;

    @Override
    public void run() {
        try {
            server.start();
        } catch (RuntimeException e) {
            exception = e;
        } catch (Error e) {
            error = e;
        }
    }

    void rethrow() {
        if (exception != null) throw exception;
        if (error != null) throw error;
    }
}
StartTask task = new StartTask();
Thread thread = CustomThreadFactory.defaultThreadFactory(daemon).newThread(task);
thread.start();
try {
    thread.join();
} catch (InterruptedException e) {
    //should not happen. but propagate interrupted flag just in case.
    Thread.currentThread().interrupt();
}
task.rethrow();
``` 

i decided to use tool that JDK provides for this – FutureTask

Reply:
I don't see why the current implementation has to do so much exception handling, I would expect it to have to do none.

This also all runs in the main thread.

Reply:
In the current implementation you just call server.start() in main thread. It can complete successfully or with Error or RuntimeException. In the latest case the HTTPServer will not be created and exception will be propagated to HTTPServer constructor caller. Now put server.start() into a separate thread and i want to propagate potential Error/RuntimeException to the HTTPServer again.

Reply:
I must be missing something here. All I see is rethrowing exceptions in the same thread. Thus this seems quite a bit more complicated than it needs to be.

Reply:
Ok. Will try to explain. RuntimeException or Error may happen in a thread that i launch in 
```java
DaemonThreadFactory.defaultThreadFactory(daemon).newThread(startTask).start();
```
The call `startTask.get()` will block until task is completed. Since it is blocking call it can be interrupted and throw InterruptedException. InterruptedException is a checked exception and should be handled explicitly. The same with an ExecutionException which is in signature of a `Future.get` and I have to handle it as well. Since I can not extend signature of the HTTPServer constructors via new checked exception, I have handle it here and the best way to do not swallow an error is to re-throw as  RuntimeException/Error.

Reply:
@brian-brazil any thoughts?